apps/homework/views.py

```diff
@@ -54,53 +54,159 @@ def get_context_data(self, **kwargs):
+            context['submission'] = submission
+            context['definition'] = submission.definition
+        except ObjectDoesNotExist:
+            raise Http404('Definition object not found')
+        return context
+
+    def test_func(self):
+        submission_pk = self.kwargs.get('submission_pk')
+        try:
+            submission = Submission.objects.get(pk=submission_pk)
+        except ObjectDoesNotExist:
+            raise Http404('Definition object not found')
+        if self.request.user.is_superuser:
+            return True
+        if self.request.user.instructor:
+            if self.request.user.instructor == submission.definition.klass.instructor:
+                return True
+        try:
+            if self.request.user.klass_memberships.get(klass__pk=submission.definition.klass.pk):
+                return True
+        except ObjectDoesNotExist:
+            return False
+        return False
+
+class SubmissionListView(LoginRequiredMixin, UserPassesTestMixin, TemplateView):
+    template_name = 'homework/submission_list.html'
+
+    def get_context_data(self, **kwargs):
+        context = super(SubmissionListView, self).get_context_data(**kwargs)
+        definition_pk = self.kwargs.get('definition_pk')
+        submissions = None
+        try:
+            definition = Definition.objects.get(pk=definition_pk)
+            context['definition'] = definition
+            context['klass'] = definition.klass
+        except ObjectDoesNotExist:
+            raise Http404('Definition object not found')
+
+        # if user is instructor of the class
+        if self.request.user.instructor and self.request.user.instructor == definition.klass.instructor:
+            try:
+                submissions = Submission.objects.filter(definition=definition)
+                context['instructor_view'] = True
+            except ObjectDoesNotExist:
+                raise Http404('Submission object not found')
+
+        # if user is member of the class
+        else:
+            try:
+                klass_membership = self.request.user.klass_memberships.get(klass__pk=definition.klass.pk)
+            except ObjectDoesNotExist:
+                raise Http404('Student Membership object not found')
+            try:
+                if definition.team_based:
+                    team = klass_membership.team
+                    submissions = Submission.objects.filter(definition=definition, team=team)
+                    context['team'] = team
+                else:
+                    submissions = Submission.objects.filter(definition=definition, creator=klass_membership)
+            except ObjectDoesNotExist:
+                raise Http404('Submission object not found')
+        context['submissions'] = submissions
+        context['user'] = self.request.user
+        return context
+
+    def test_func(self):
+        definition_pk = self.kwargs.get('definition_pk')
```

Comment:
If context is populated by this point, and we can refer to it through self then it may be best to grab it this way so we're not generating more queries. (For all of the views using `test_func`)

Reply:
Would it be possible to do this if the order of inheritance was changed? The TemplateView base class will be the one to call `get_context_data`.